crypto/evp/e_aes.c

```diff
@@ -1581,20 +1663,483 @@ static int s390x_aes_xts_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
 static int s390x_aes_xts_ctrl(EVP_CIPHER_CTX *, int type, int arg, void *ptr);
 # define s390x_aes_xts_cleanup aes_xts_cleanup
 
-# define S390X_AES_CCM_CTX		EVP_AES_CCM_CTX
-# define S390X_aes_128_ccm_CAPABLE	0
-# define S390X_aes_192_ccm_CAPABLE	0
-# define S390X_aes_256_ccm_CAPABLE	0
+# define S390X_aes_128_ccm_CAPABLE (S390X_aes_128_CAPABLE &&		\
+                                    (OPENSSL_s390xcap_P.kmac[0] &	\
+                                     S390X_CAPBIT(S390X_AES_128)))
+# define S390X_aes_192_ccm_CAPABLE (S390X_aes_192_CAPABLE &&		\
+                                    (OPENSSL_s390xcap_P.kmac[0] &	\
+                                     S390X_CAPBIT(S390X_AES_192)))
+# define S390X_aes_256_ccm_CAPABLE (S390X_aes_256_CAPABLE &&		\
+                                    (OPENSSL_s390xcap_P.kmac[0] &	\
+                                     S390X_CAPBIT(S390X_AES_256)))
+
+# define S390X_CCM_AAD_FLAG	0x40
+
+/*-
+ * Set nonce and length fields. Code is big-endian.
+ */
+static inline void s390x_aes_ccm_setiv(S390X_AES_CCM_CTX *ctx,
+                                          const unsigned char *nonce,
+                                          size_t mlen)
+{
+    ctx->aes.ccm.nonce.b[0] &= ~S390X_CCM_AAD_FLAG;
+    ctx->aes.ccm.nonce.g[1] = mlen;
+    memcpy(ctx->aes.ccm.nonce.b + 1, nonce, 15 - ctx->aes.ccm.l);
+}
+
+/*-
+ * Process additional authenticated data. Code is big-endian.
+ */
+static void s390x_aes_ccm_aad(S390X_AES_CCM_CTX *ctx, const unsigned char *aad,
+                              size_t alen)
+{
+    unsigned char *ptr;
+    int i, rem;
+
+    if (!alen)
+        return;
+
+    ctx->aes.ccm.nonce.b[0] |= S390X_CCM_AAD_FLAG;
+
+    /* Suppress 'type-punned pointer dereference' warning. */
+    ptr = ctx->aes.ccm.buf.b;
+
+    if (alen < ((1 << 16) - (1 << 8))) {
+        *(uint16_t *)ptr = alen;
+        i = 2;
+    } else if ((unsigned long long)alen < 1ULL << 32) {
```

Comment:
I get warning in 32-bit build with --strict-warnings here. In modes/ccm128.c this is resolved effectively as
```
if (sizeof(alen) == 8 && alen < ((size_t)1 << (32 % (sizeof(alen) * 8))
```
However, it seems that `if (sizeof(alen) == 8 && alen < (size_t)1 << 32)` works too. I mean compiler is smart enough to figure out that second condition is never evaluated and doesn't complain about shift by 32 being out of bounds...

Reply:
Though of course above suggested condition would be dead-wrong in the context. I mean it would force all inputs longer than 65280 to be treated as longer than 2GB on 32-bit application context. So do *not* use suggested `if`, but fix the warning. By using logic similar to one in modes/ccm128.c.

Reply:
yes im aware of how its done in ccm128.c but i thought i could simplify this, assuming compiler would just optimize it i.e., an silently turn an else-if that is always true into an else. Seems i somehow i forgot to check strict-warnings.
Anyway, i think the ccm128.c code could need a comment like "supress warning X on 32 bit".